Compile failure: radare2 4.4.0 does not compile on MacOS 10.9 Mavericks

I am running MacOS 10.9. Yes, I am aware that it is no longer officially supported, but I am still running it on this machine, and I would like to run the very nice looking rasm2 on it as well.

When running `sys/install.sh` as instructed in the readme (as well as when trying to install via the last supported version of brew), compilation fails because the project uses `clock_gettime` and `CLOCK_MONOTONIC`, which are not supported in MacOS 10.11 and earlier, without using any compatibility layer such as [PosixMachTiming](https://github.com/pineapplemachine/PosixMachTiming/tree/improve-degenerate-case-behavior).

It would be fantastic if such a compatibility layer could be added so that radare2 can be compiled on MacOS 10.9.

As a workaround, though, I was able to get 3.1.3 installed and apparently running fine. I don't suppose anyone can let me know what was the last supported version on MacOS 10.9? I could not find anything in the release changelogs that indicated dropping support for older versions of MacOS.

Thanks!

### Work environment

| Questions                                            | Answers
|------------------------------------------------------|--------------------
| OS/arch/bits (mandatory)                             | OS X 10.9.5, x86-64
| File format of the file you reverse (mandatory)      | N/A
| Architecture/bits of the file (mandatory)            | N/A
| r2 -v full output, **not truncated** (mandatory)     | N/A

### Expected behavior

The build should complete successfully on MacOS 10.9 Mavericks.

### Actual behavior

The build fails on MacOS 10.9 Mavericks.

### Steps to reproduce the behavior 

Attempt to build radare2 4.4.0 on MacOS 10.11 or earlier. 

This fails because clock_gettime and CLOCK_MONOTONIC were not supported until MacOS 10.12; to make these available on MacOS 10.11 or earlier, a compatibility layer such as [this one](https://github.com/pineapplemachine/PosixMachTiming/tree/improve-degenerate-case-behavior) can be used.

- Please share the binary if it is shareable by drag and dropping it here in a zip archive (mandatory)

N/A

- Use asciinema to describe the issue and share the link here (mandatory if u can't share the binary) 

N/A

- Use code markdown `CODE` to make your code visible

`N/A`

- Or even better, create a Pull-Request containing the test case in the test/ folder. See for example test/new/db/cmd/cmd_search.

N/A

### Additional Logs, screenshots, source-code,  configuration dump, ...

Here is the relevant part of the build log:

``` text
CC run.c
CC r2r.c
clang: warning: argument unused during compilation: '-pie'
clang: warning: argument unused during compilation: '-pie'
clang: warning: argument unused during compilation: '-pie'
run.c:640:3: warning: implicit declaration of function 'clock_gettime' is invalid in C99
      [-Wimplicit-function-declaration]
                clock_gettime (CLOCK_MONOTONIC, &timeout_abs);
                ^
run.c:640:18: error: use of undeclared identifier 'CLOCK_MONOTONIC'
                clock_gettime (CLOCK_MONOTONIC, &timeout_abs);
                               ^
run.c:678:19: error: use of undeclared identifier 'CLOCK_MONOTONIC'
                        clock_gettime (CLOCK_MONOTONIC, &now);
                                       ^
1 warning and 2 errors generated.
make[2]: *** [run.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [all] Error 1
make: *** [all] Error 2
```

@ret2libc Cloning the current master branch and attempting to compile produces the same error as before:

``` text
CC run.c
CC r2r.c
clang: warning: argument unused during compilation: '-pie'
clang: warning: argument unused during compilation: '-pie'
clang: warning: argument unused during compilation: '-pie'
run.c:640:3: warning: implicit declaration of function 'clock_gettime' is invalid in
      C99 [-Wimplicit-function-declaration]
                clock_gettime (CLOCK_MONOTONIC, &timeout_abs);
                ^
run.c:640:18: error: use of undeclared identifier 'CLOCK_MONOTONIC'
                clock_gettime (CLOCK_MONOTONIC, &timeout_abs);
                               ^
run.c:678:19: error: use of undeclared identifier 'CLOCK_MONOTONIC'
                        clock_gettime (CLOCK_MONOTONIC, &now);
                                       ^
1 warning and 2 errors generated.
make[2]: *** [run.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [all] Error 1
make: *** [all] Error 2
```

@ret2libc This resulted in a similar error at a different point in the build process.

I ran `./configure --without-r2r` and then got this error running `make`:

``` text
gcc -c  -MD -fno-common   -fPIC -g -Wall -D__UNIX__=1 -I/Users/pineapple/apps/radare2-4.4.0/libr -I/Users/pineapple/apps/radare2-4.4.0/libr/include -fvisibility=hidden -o rabin2.o rabin2.c
gcc -c  -MD -fno-common   -fPIC -g -Wall -D__UNIX__=1 -I/Users/pineapple/apps/radare2-4.4.0/libr -I/Users/pineapple/apps/radare2-4.4.0/libr/include -fvisibility=hidden -o rafind2.o rafind2.c
gcc -c  -MD -fno-common   -fPIC -g -Wall -D__UNIX__=1 -I/Users/pineapple/apps/radare2-4.4.0/libr -I/Users/pineapple/apps/radare2-4.4.0/libr/include -fvisibility=hidden -o r2agent.o r2agent.c
gcc -c  -MD -fno-common   -fPIC -g -Wall -D__UNIX__=1 -I/Users/pineapple/apps/radare2-4.4.0/libr -I/Users/pineapple/apps/radare2-4.4.0/libr/include -fvisibility=hidden -o radiff2.o radiff2.c
gcc -c  -MD -fno-common   -fPIC -g -Wall -D__UNIX__=1 -I/Users/pineapple/apps/radare2-4.4.0/libr -I/Users/pineapple/apps/radare2-4.4.0/libr/include -fvisibility=hidden -o radare2.o radare2.c
gcc -c  -MD -fno-common   -fPIC -g -Wall -D__UNIX__=1 -I/Users/pineapple/apps/radare2-4.4.0/libr -I/Users/pineapple/apps/radare2-4.4.0/libr/include -fvisibility=hidden -o rahash2.o rahash2.c
"gcc -dynamiclib -o libr_main.dylib -Wl,-install_name,/usr/local/lib/libr_main.dylib main.o rax2.o rasm2.o ragg2.o rarun2.o rabin2.o rafind2.o r2agent.o radiff2.o radare2.o rahash2.o  -L/Users/pineapple/apps/radare2-4.4.0/libr/config -L/Users/pineapple/apps/radare2-4.4.0/libr/cons -L/Users/pineapple/apps/radare2-4.4.0/libr/io -L/Users/pineapple/apps/radare2-4.4.0/libr/util -L/Users/pineapple/apps/radare2-4.4.0/libr/flag -L/Users/pineapple/apps/radare2-4.4.0/libr/asm -L/Users/pineapple/apps/radare2-4.4.0/libr/core -L/Users/pineapple/apps/radare2-4.4.0/libr/debug -L/Users/pineapple/apps/radare2-4.4.0/libr/hash -L/Users/pineapple/apps/radare2-4.4.0/libr/bin -L/Users/pineapple/apps/radare2-4.4.0/libr/lang -L/Users/pineapple/apps/radare2-4.4.0/libr/io -L/Users/pineapple/apps/radare2-4.4.0/libr/anal -L/Users/pineapple/apps/radare2-4.4.0/libr/parse -L/Users/pineapple/apps/radare2-4.4.0/libr/bp -L/Users/pineapple/apps/radare2-4.4.0/libr/egg -L/Users/pineapple/apps/radare2-4.4.0/libr/reg -L/Users/pineapple/apps/radare2-4.4.0/libr/search -L/Users/pineapple/apps/radare2-4.4.0/libr/syscall -L/Users/pineapple/apps/radare2-4.4.0/libr/socket -L/Users/pineapple/apps/radare2-4.4.0/libr/fs -L/Users/pineapple/apps/radare2-4.4.0/libr/magic -L/Users/pineapple/apps/radare2-4.4.0/libr/crypto -lr_config -lr_cons -lr_io -lr_util -lr_flag -lr_asm -lr_core -lr_debug -lr_hash -lr_bin -lr_lang -lr_io -lr_anal -lr_parse -lr_bp -lr_egg -lr_reg -lr_search -lr_syscall -lr_socket -lr_fs -lr_magic -lr_crypto   -fPIC -fvisibility=hidden    -fvisibility=hidden"
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C binr
make[2]: Nothing to be done for `all'.
make[2]: Nothing to be done for `all'.
make[2]: Nothing to be done for `all'.
make[2]: Nothing to be done for `all'.
make[2]: Nothing to be done for `all'.
make[2]: Nothing to be done for `all'.
make[2]: Nothing to be done for `all'.
make[2]: Nothing to be done for `all'.
make[2]: Nothing to be done for `all'.
make[2]: Nothing to be done for `all'.
gcc -c  -MD -fno-common   -fPIC -g -Wall -D__UNIX__=1 -I/Users/pineapple/apps/radare2-4.4.0/libr/../shlr/zip/include -pie -I/Users/pineapple/apps/radare2-4.4.0/libr/..//libr/include -o run.o run.c
clang: warning: argument unused during compilation: '-pie'
run.c:640:3: warning: implicit declaration of function 'clock_gettime' is invalid in C99 [-Wimplicit-function-declaration]
                clock_gettime (CLOCK_MONOTONIC, &timeout_abs);
                ^
run.c:640:18: error: use of undeclared identifier 'CLOCK_MONOTONIC'
                clock_gettime (CLOCK_MONOTONIC, &timeout_abs);
                               ^
run.c:678:19: error: use of undeclared identifier 'CLOCK_MONOTONIC'
                        clock_gettime (CLOCK_MONOTONIC, &now);
```

--- a/binr/r2r/run.c
+++ b/binr/r2r/run.c
@@ -2,12 +2,6 @@
 
 #include "r2r.h"
 
-#define NSEC_PER_SEC  1000000000
-#define NSEC_PER_MSEC 1000000
-#define USEC_PER_SEC  1000000
-#define NSEC_PER_USEC 1000
-#define USEC_PER_MSEC 1000
-
 #if __WINDOWS__
 struct r2r_subprocess_t {
 	HANDLE stdin_write;
@@ -236,20 +230,6 @@ error:
 	goto beach;
 }
 
-static ut64 now_us(void) {
-	LARGE_INTEGER f;
-	if (!QueryPerformanceFrequency (&f)) {
-		return 0;
-	}
-	LARGE_INTEGER v;
-	if (!QueryPerformanceCounter (&v)) {
-		return 0;
-	}
-	v.QuadPart *= 1000000;
-	v.QuadPart /= f.QuadPart;
-	return v.QuadPart;
-}
-
 R_API bool r2r_subprocess_wait(R2RSubprocess *proc, ut64 timeout_ms) {
 	OVERLAPPED stdout_overlapped = { 0 };
 	stdout_overlapped.hEvent = CreateEvent (NULL, TRUE, FALSE, NULL);
@@ -265,7 +245,7 @@ R_API bool r2r_subprocess_wait(R2RSubprocess *proc, ut64 timeout_ms) {
 
 	ut64 timeout_us_abs = UT64_MAX;
 	if (timeout_ms != UT64_MAX) {
-		timeout_us_abs = now_us () + timeout_ms * USEC_PER_MSEC;
+		timeout_us_abs = r_sys_now_mono () + timeout_ms * USEC_PER_MSEC;
 	}
 
 	ut8 stdout_buf[0x500];
@@ -307,7 +287,7 @@ R_API bool r2r_subprocess_wait(R2RSubprocess *proc, ut64 timeout_ms) {
 		
 		DWORD timeout = INFINITE;
 		if (timeout_us_abs != UT64_MAX) {
-			ut64 now = now_us ();
+			ut64 now = r_sys_now_mono ();
 			if (now >= timeout_us_abs) {
 				return false;
 			}
@@ -635,12 +615,9 @@ error:
 }
 
 R_API bool r2r_subprocess_wait(R2RSubprocess *proc, ut64 timeout_ms) {
-	struct timespec timeout_abs;
+	ut64 timeout_abs;
 	if (timeout_ms != UT64_MAX) {
-		clock_gettime (CLOCK_MONOTONIC, &timeout_abs);
-		timeout_abs.tv_nsec += timeout_ms * NSEC_PER_MSEC;
-		timeout_abs.tv_sec += timeout_abs.tv_nsec / NSEC_PER_SEC;
-		timeout_abs.tv_nsec = timeout_abs.tv_nsec % NSEC_PER_SEC;
+		timeout_abs = r_sys_now_mono () + timeout_ms * R_USEC_PER_MSEC;
 	}
 
 	int r = 0;
@@ -674,15 +651,12 @@ R_API bool r2r_subprocess_wait(R2RSubprocess *proc, ut64 timeout_ms) {
 		struct timeval timeout_s;
 		struct timeval *timeout = NULL;
 		if (timeout_ms != UT64_MAX) {
-			struct timespec now;
-			clock_gettime (CLOCK_MONOTONIC, &now);
-			st64 usec_diff = ((st64)timeout_abs.tv_sec - now.tv_sec) * USEC_PER_SEC
-					+ ((st64)timeout_abs.tv_nsec - now.tv_nsec) / NSEC_PER_USEC;
+			st64 usec_diff = timeout_abs - r_sys_now_mono ();
 			if (usec_diff <= 0) {
 				break;
 			}
-			timeout_s.tv_sec = usec_diff / USEC_PER_SEC;
-			timeout_s.tv_usec = usec_diff % USEC_PER_SEC;
+			timeout_s.tv_sec = usec_diff / R_USEC_PER_SEC;
+			timeout_s.tv_usec = usec_diff % R_USEC_PER_SEC;
 			timeout = &timeout_s;
 		}
 		r = select (nfds, &rfds, NULL, NULL, timeout);

--- a/binr/r2r/run.c
+++ b/binr/r2r/run.c
@@ -245,7 +245,7 @@ R_API bool r2r_subprocess_wait(R2RSubprocess *proc, ut64 timeout_ms) {
 
 	ut64 timeout_us_abs = UT64_MAX;
 	if (timeout_ms != UT64_MAX) {
-		timeout_us_abs = r_sys_now_mono () + timeout_ms * USEC_PER_MSEC;
+		timeout_us_abs = r_sys_now_mono () + timeout_ms * R_USEC_PER_MSEC;
 	}
 
 	ut8 stdout_buf[0x500];
@@ -291,7 +291,7 @@ R_API bool r2r_subprocess_wait(R2RSubprocess *proc, ut64 timeout_ms) {
 			if (now >= timeout_us_abs) {
 				return false;
 			}
-			timeout = (DWORD)((timeout_us_abs - now) / USEC_PER_MSEC);
+			timeout = (DWORD)((timeout_us_abs - now) / R_USEC_PER_MSEC);
 		}
 		DWORD signaled = WaitForMultipleObjects (handles.len, handles.a, FALSE, timeout);
 		if (!stdout_eof && signaled == stdout_index) {

--- a/binr/r2r/run.c
+++ b/binr/r2r/run.c
@@ -651,10 +651,11 @@ R_API bool r2r_subprocess_wait(R2RSubprocess *proc, ut64 timeout_ms) {
 		struct timeval timeout_s;
 		struct timeval *timeout = NULL;
 		if (timeout_ms != UT64_MAX) {
-			st64 usec_diff = timeout_abs - r_sys_now_mono ();
-			if (usec_diff <= 0) {
+			ut64 now = r_sys_now_mono ();
+			if (now >= timeout_abs) {
 				break;
 			}
+			ut64 usec_diff = timeout_abs - r_sys_now_mono ();
 			timeout_s.tv_sec = usec_diff / R_USEC_PER_SEC;
 			timeout_s.tv_usec = usec_diff % R_USEC_PER_SEC;
 			timeout = &timeout_s;

--- a/libr/include/r_util/r_sys.h
+++ b/libr/include/r_util/r_sys.h
@@ -37,7 +37,19 @@ R_API int r_sys_signal(int sig, void (*handler) (int));
 R_API void r_sys_env_init(void);
 R_API char **r_sys_get_environ(void);
 R_API void r_sys_set_environ(char **e);
+
+#define R_NSEC_PER_SEC  1000000000
+#define R_NSEC_PER_MSEC 1000000
+#define R_USEC_PER_SEC  1000000
+#define R_NSEC_PER_USEC 1000
+#define R_USEC_PER_MSEC 1000
+
+// wall clock time in microseconds
 R_API ut64 r_sys_now(void);
+
+// monotonic time in microseconds
+R_API ut64 r_sys_now_mono(void);
+
 R_API const char *r_time_to_string (ut64 ts);
 R_API int r_sys_fork(void);
 // nocleanup = false => exit(); true => _exit()

--- a/libr/util/sys.c
+++ b/libr/util/sys.c
@@ -2,6 +2,8 @@
 
 #if __linux__
 #include <time.h>
+#elif __APPLE__ && !defined(MAC_OS_X_VERSION_10_12)
+#include <mach/mach_time.h>
 #endif
 
 #include <r_userconf.h>
@@ -231,6 +233,11 @@ R_API ut64 r_sys_now_mono(void) {
 	v.QuadPart *= 1000000;
 	v.QuadPart /= f.QuadPart;
 	return v.QuadPart;
+#elif __APPLE__ && !defined(MAC_OS_X_VERSION_10_12)
+	ut64 ticks = mach_absolute_time ();
+	static mach_timebase_info_data_t tb;
+	mach_timebase_info (&tb);
+	return ((ticks * tb.numer) / tb.denom) / R_NSEC_PER_USEC;
 #else
 	struct timespec now;
 	clock_gettime (CLOCK_MONOTONIC, &now);

--- a/libr/util/sys.c
+++ b/libr/util/sys.c
@@ -218,6 +218,27 @@ R_API ut64 r_sys_now(void) {
 	return ret;
 }
 
+R_API ut64 r_sys_now_mono(void) {
+#if __WINDOWS__
+	LARGE_INTEGER f;
+	if (!QueryPerformanceFrequency (&f)) {
+		return 0;
+	}
+	LARGE_INTEGER v;
+	if (!QueryPerformanceCounter (&v)) {
+		return 0;
+	}
+	v.QuadPart *= 1000000;
+	v.QuadPart /= f.QuadPart;
+	return v.QuadPart;
+#else
+	struct timespec now;
+	clock_gettime (CLOCK_MONOTONIC, &now);
+	return now.tv_sec * R_USEC_PER_SEC
+		+ now.tv_nsec / R_NSEC_PER_USEC;
+#endif
+}
+
 R_API int r_sys_truncate(const char *file, int sz) {
 #if __WINDOWS__
 	int fd = r_sandbox_open (file, O_RDWR, 0644);

